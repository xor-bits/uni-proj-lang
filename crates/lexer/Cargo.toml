[package]
name = "lexer"
version.workspace = true
edition.workspace = true

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
# prob faster lexing + less boilerplate:
# logos = "0.13"

# only for unit tests
[dev-dependencies]
serde = { version = "1.0", features = ["derive"] }
insta = { version = "1.40", features = ["yaml"] }
