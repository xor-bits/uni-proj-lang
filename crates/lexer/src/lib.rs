#[cfg(test)]
use serde::{ser::SerializeStruct, Serialize};
use std::{fmt, ops::Range};

//

macro_rules! some {
    ($v:expr) => {
        if let Some(v) = $v {
            return Some(v);
        }
    };
}

macro_rules! exhaustive {
    ($($tok:ident::$var:ident $({ $($t:tt)* })?,)*) => {{
        use crate::Token::*;

        #[allow(unreachable_code)]
        fn _assert_exhaustive() {
            match unreachable!() {
                $($var { .. } => {})*
            }
        }

        [
            $($tok::$var $({ $($t)* })?),*
        ]
    }};
}

//

pub type Result<T, E = Error> = core::result::Result<T, E>;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Error {
    UnexpectedEoi,
    ExtraTokens,
}

impl fmt::Display for Error {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            Error::UnexpectedEoi => f.write_str("unexpected end of input"),
            Error::ExtraTokens => f.write_str("found extra tokens"),
        }
    }
}

//

#[cfg_attr(test, derive(Serialize))]
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Token {
    /// `;`
    Semi,

    /// `=`
    Assign,

    /// `(`
    LParen,

    /// `)`
    RParen,

    /// `{`
    LBrace,

    /// `}`
    RBrace,

    /// `+`
    Plus,

    /// `-`
    Minus,

    /// `*`
    Asterisk,

    /// `/`
    Slash,

    /// `%`
    Percent,

    /// `:`
    Colon,

    /// `,`
    Comma,

    /// `@`
    At,

    /// `&`
    Ampersand,

    /// `<`
    Lt,

    /// `>`
    Gt,

    /// `<=`
    Le,

    /// `>=`
    Ge,

    /// `==`
    Eq,

    /// `!=`
    Neq,

    /// `&&`
    And,

    /// `||`
    Or,

    /// `:=`
    Walrus,

    /// `->`
    RArrow,

    /// `fn`
    Fn,

    /// `for`
    For,

    /// `return`
    Return,

    /// `struct`
    Struct,

    /// `test`
    Test,

    /// single line comments like `// this is a comment`
    LineComment,

    /// identifiers like `_fn_name`
    Ident,

    /// floating point literal like `5.2`
    LitFloat,

    /// integer literal like `0xFFu8`
    LitInt,

    /// string literal like `"test"`
    LitStr,

    /// end of input
    Eoi,
}

impl Token {
    pub const fn enumerate() -> &'static [Self] {
        &exhaustive![
            Token::Semi,
            Token::Assign,
            Token::LParen,
            Token::RParen,
            Token::LBrace,
            Token::RBrace,
            Token::Plus,
            Token::Minus,
            Token::Asterisk,
            Token::Slash,
            Token::Percent,
            Token::Colon,
            Token::Comma,
            Token::At,
            Token::Ampersand,
            Token::Lt,
            Token::Gt,
            Token::Le,
            Token::Ge,
            Token::Eq,
            Token::Neq,
            Token::And,
            Token::Or,
            Token::Walrus,
            Token::RArrow,
            Token::Fn,
            Token::For,
            Token::Return,
            Token::Struct,
            Token::Test,
            Token::LineComment,
            Token::Ident,
            Token::LitFloat,
            Token::LitInt,
            Token::LitStr,
            Token::Eoi,
        ]
    }

    pub const fn token_type(self) -> TokenType {
        match self {
            Token::Semi => TokenType::Symbols(";"),
            Token::Assign => TokenType::Symbols("="),
            Token::LParen => TokenType::Symbols("("),
            Token::RParen => TokenType::Symbols(")"),
            Token::LBrace => TokenType::Symbols("{"),
            Token::RBrace => TokenType::Symbols("}"),
            Token::Plus => TokenType::Symbols("+"),
            Token::Minus => TokenType::Symbols("-"),
            Token::Asterisk => TokenType::Symbols("*"),
            Token::Slash => TokenType::Symbols("/"),
            Token::Percent => TokenType::Symbols("%"),
            Token::Colon => TokenType::Symbols(":"),
            Token::Comma => TokenType::Symbols(","),
            Token::At => TokenType::Symbols("@"),
            Token::Ampersand => TokenType::Symbols("&"),
            Token::Lt => TokenType::Symbols("<"),
            Token::Gt => TokenType::Symbols(">"),
            Token::Le => TokenType::Symbols("<="),
            Token::Ge => TokenType::Symbols(">="),
            Token::Eq => TokenType::Symbols("=="),
            Token::Neq => TokenType::Symbols("!="),
            Token::And => TokenType::Symbols("&&"),
            Token::Or => TokenType::Symbols("||"),
            Token::Walrus => TokenType::Symbols(":="),
            Token::RArrow => TokenType::Symbols("->"),
            Token::Fn => TokenType::Keyword("fn"),
            Token::For => TokenType::Keyword("for"),
            Token::Return => TokenType::Keyword("return"),
            Token::Struct => TokenType::Keyword("struct"),
            Token::Test => TokenType::Keyword("test"),
            Token::LineComment => TokenType::Other,
            Token::Ident => TokenType::Other,
            Token::LitFloat => TokenType::Other,
            Token::LitInt => TokenType::Other,
            Token::LitStr => TokenType::Other,
            Token::Eoi => TokenType::Other,
        }
    }
}

impl fmt::Display for Token {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self.token_type() {
            TokenType::Symbols(s) => f.write_str(s),
            TokenType::Keyword(s) => f.write_str(s),
            TokenType::Other => write!(f, "{self:?}"),
        }
    }
}

//

#[derive(Debug, Clone, Copy)]
pub enum TokenType {
    Symbols(&'static str),
    Keyword(&'static str),
    Other,
}

//

#[cfg_attr(test, derive(Serialize))]
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct SpannedToken<'a> {
    token: Token,
    span: Span<'a>,
}

impl<'a> SpannedToken<'a> {
    pub const fn from_whole_str(token: Token, str: &'a str) -> Self {
        Self {
            token,
            span: Span::from_whole_str(str),
        }
    }

    pub fn token(&self) -> Token {
        self.token
    }

    pub fn span(&self) -> Range<usize> {
        self.span.span()
    }

    pub fn as_str(&self) -> &'a str {
        self.span.as_str()
    }
}

impl fmt::Display for SpannedToken<'_> {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        if self.token == Token::Eoi {
            write!(f, "<eoi>")
        } else {
            f.write_str(self.as_str())
        }
    }
}

//

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct Span<'a> {
    from: usize,
    to: usize,
    source: &'a str,
}

impl<'a> Span<'a> {
    pub const fn from_whole_str(str: &'a str) -> Self {
        Self {
            from: 0,
            to: str.len(),
            source: str,
        }
    }

    pub fn span(&self) -> Range<usize> {
        self.from..self.to
    }

    pub fn as_str(&self) -> &'a str {
        &self.source[self.span()]
    }
}

#[cfg(test)]
impl Serialize for Span<'_> {
    fn serialize<S>(&self, serializer: S) -> core::result::Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        let mut s = serializer.serialize_struct("SpannedToken", 3)?;
        s.serialize_field("range", &format!("{:?}", self.span()))?;
        s.serialize_field("str", &self.as_str())?;
        s.end()
    }
}

//

#[derive(Debug, Clone)]
pub struct Lexer<'a> {
    source: &'a str,
    at: &'a str,
    eoi: bool,
    err: bool,
}

impl<'a> Lexer<'a> {
    pub const fn new(source: &'a str) -> Self {
        Self {
            source,
            at: source,
            eoi: false,
            err: false,
        }
    }

    // pub const fn error(&self) -> Result<()> {
    //     self.err
    // }

    fn src_index(&self) -> usize {
        self.at.as_ptr() as usize - self.source.as_ptr() as usize
    }

    fn advance_by(&mut self, len: usize) {
        self.at = &self.at[len..];
    }

    fn advance(&mut self, span: &Span) {
        self.advance_by(span.span().len())
    }

    fn span_to(&self, len: usize) -> Span<'a> {
        let at = self.src_index();
        Span {
            from: at,
            to: at + len,
            source: self.source,
        }
    }

    fn spanned_token_to(&mut self, token: Token, len: usize) -> Option<Result<SpannedToken<'a>>> {
        let span = self.span_to(len);
        self.advance(&span);
        Some(Ok(SpannedToken { token, span }))
    }

    fn try_match(&mut self, token: Token) -> Option<Result<SpannedToken<'a>>> {
        let TokenType::Symbols(str) = token.token_type() else {
            return None;
        };
        if !self.at.starts_with(str) {
            return None;
        }

        self.spanned_token_to(token, str.len())
    }

    fn try_match_number(&mut self) -> Option<Result<SpannedToken<'a>>> {
        let first = self.at.chars().next()?;

        if !first.is_ascii_digit() {
            return None;
        }

        let mut dot_found = false;
        let term = self.at[1..]
            .find(|c: char| {
                !c.is_ascii_digit()
                    || if dot_found {
                        false
                    } else if c == '.' {
                        dot_found = true;
                        true
                    } else {
                        false
                    }
            })
            .map(|term| term + 1)
            .unwrap_or(self.at.len());

        let token = if dot_found {
            Token::LitFloat
        } else {
            Token::LitInt
            // let num = self.at[..term].parse::<i128>().unwrap(); // the unwrap should never fail
            // Token::LitInt(num)
        };

        self.spanned_token_to(token, term)
    }

    fn try_match_str(&mut self) -> Option<Result<SpannedToken<'a>>> {
        if !self.at.starts_with('"') {
            return None;
        }

        let Some(term) = self.at[1..].find('"') else {
            return self.err(Error::UnexpectedEoi);
        };

        self.spanned_token_to(Token::LitStr, term + 2)
    }

    fn try_match_ident(&mut self) -> Option<Result<SpannedToken<'a>>> {
        let first = self.at.chars().next()?;

        if !(first == '_' || first.is_alphabetic()) {
            return None;
        }

        let term = self.at[1..]
            .find(|c: char| !(c.is_alphanumeric() || c == '_'))
            .map(|term| term + 1)
            .unwrap_or(self.at.len());

        self.spanned_token_to(Token::Ident, term)
    }

    fn try_match_comment(&mut self) -> Option<Result<SpannedToken<'a>>> {
        if !self.at.starts_with("//") {
            return None;
        }

        let term = self.at[2..]
            .lines()
            .next()
            .map(|line| line.len() + 2)
            .unwrap_or(self.at.len());

        self.spanned_token_to(Token::LineComment, term)
    }

    fn err(&mut self, err: Error) -> Option<Result<SpannedToken<'a>>> {
        self.eoi = true;
        Some(Err(err))
    }

    fn eoi(&mut self) -> Option<Result<SpannedToken<'a>>> {
        self.eoi = true;
        self.spanned_token_to(Token::Eoi, 0)
    }
}

impl<'a> Iterator for Lexer<'a> {
    type Item = Result<SpannedToken<'a>>;

    fn next(&mut self) -> Option<Self::Item> {
        if self.eoi {
            return None;
        }
        if self.err {
            return self.eoi();
        }
        self.at = self.at.trim();

        // tokens in the order of 'more specific -> less specific' to emulate priority
        // for example := should be parsed as Init instead of Type and Assign

        // custom tokens
        some!(self.try_match_str());
        some!(self.try_match_comment());
        some!(self.try_match_number());

        // simple const str tokens
        for (_, token) in Token::enumerate()
            .iter()
            .rev()
            .copied()
            .filter_map(|tok| match tok.token_type() {
                TokenType::Symbols(s) => Some((tok, s)),
                _ => None,
            })
            // just a debug assert to make sure that the tokens are in the correct order
            .scan(usize::MAX, |prev, (now, s)| {
                debug_assert!(*prev >= s.len());
                Some((s.len(), now))
            })
        {
            some!(self.try_match(token));
        }

        // idents after keywords
        if let Some(ident) = self.try_match_ident() {
            let ident = match ident {
                Ok(ident) => ident,
                Err(err) => return Some(Err(err)),
            };

            for (s, token) in
                Token::enumerate()
                    .iter()
                    .rev()
                    .copied()
                    .filter_map(|tok| match tok.token_type() {
                        TokenType::Keyword(s) => Some((s, tok)),
                        _ => None,
                    })
            {
                if ident.as_str() == s {
                    return Some(Ok(SpannedToken {
                        token,
                        span: ident.span,
                    }));
                }
            }

            return Some(Ok(ident));
        }

        if !self.at.is_empty() {
            return self.err(Error::ExtraTokens);
        }

        self.eoi()
    }
}

//

#[cfg(test)]
mod tests {
    use crate::{Lexer, Result, SpannedToken, Token, TokenType};

    use insta::assert_yaml_snapshot;
    use serde::Serialize;

    fn lex(str: &str) -> impl Serialize + '_ {
        let tokens: Result<Vec<SpannedToken>> = Lexer::new(str).collect();
        (str, tokens.map_err(|err| err.to_string()))
    }

    #[test]
    fn lex_test() {
        assert_yaml_snapshot!(lex(
            "in->\"vali\"@->:d synt(ax} {but) t;h:e l=e:=xer should handle it"
        ));
    }

    #[test]
    fn lex_str_lit() {
        assert_yaml_snapshot!(lex("_a2"));
        assert_yaml_snapshot!(lex(""));
        assert_yaml_snapshot!(lex("\""));
        assert_yaml_snapshot!(lex("\"a"));
        assert_yaml_snapshot!(lex("\"a\""));
    }

    #[test]
    fn lex_all_simple() {
        let all_tokens: String = Token::enumerate()
            .iter()
            .filter_map(|tok| match tok.token_type() {
                TokenType::Symbols(s) | TokenType::Keyword(s) => Some(format!("{s} ")),
                TokenType::Other => None,
            })
            .collect();

        assert_yaml_snapshot!(lex(&all_tokens));
    }

    #[test]
    fn lex_all_no_spaces() {
        let all_tokens: String = Token::enumerate()
            .iter()
            .filter_map(|tok| match tok.token_type() {
                TokenType::Symbols(s) | TokenType::Keyword(s) => Some(s),
                TokenType::Other => None,
            })
            .collect();

        assert_yaml_snapshot!(lex(&all_tokens));
    }

    #[test]
    fn lex_comment() {
        assert_yaml_snapshot!(lex("// comment\ncode"));
    }

    #[test]
    fn lex_example() {
        assert_yaml_snapshot!(lex(include_str!("../../../tests/trivial")));
    }
}
