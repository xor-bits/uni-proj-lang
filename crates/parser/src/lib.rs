use std::{collections::VecDeque, fmt, iter::Filter};

use lexer::{Lexer, SpannedToken, Token};

#[cfg(test)]
use serde::Serialize;

//

pub mod ast;

//

pub type Result<T, E = Error> = core::result::Result<T, E>;

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Error {
    Lexer(lexer::Error),
    UnexpectedToken(String),
}

impl std::error::Error for Error {
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        None
    }

    fn cause(&self) -> Option<&dyn std::error::Error> {
        self.source()
    }
}

impl fmt::Display for Error {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            Error::Lexer(err) => fmt::Display::fmt(err, f),
            Error::UnexpectedToken(tok) => write!(f, "unexpected token {tok}"),
        }
    }
}

//

#[cfg_attr(test, derive(Serialize))]
#[derive(Debug, Clone, Copy, Default)]
#[repr(u8)]
pub enum Value {
    Bool(bool),
    I32(i32),
    Str(&'static str),
    Void,
    Never,
    #[default]
    Unknown,
    Other,
}

#[cfg_attr(test, derive(Serialize))]
#[derive(Debug, Clone, Copy, Default)]
#[repr(u8)]
pub enum TypeId {
    Bool,
    I32,
    Str,
    Void,
    Never, // never matches everything
    #[default]
    Unknown,
    Other(u32),
}

impl PartialEq for TypeId {
    fn eq(&self, other: &Self) -> bool {
        match (self, other) {
            (TypeId::Bool, TypeId::Bool) => true,
            (TypeId::I32, TypeId::I32) => true,
            (TypeId::Str, TypeId::Str) => true,
            (TypeId::Void, TypeId::Void) => true,
            (TypeId::Never, _) => true,
            (_, TypeId::Never) => true,
            (TypeId::Unknown, TypeId::Unknown) => true,
            (TypeId::Other(l), TypeId::Other(r)) => l == r,
            _ => false,
        }
    }
}

impl Eq for TypeId {}

impl TypeId {
    // pub fn from_type<T: AsTypeId>() -> Self {
    //     T::TYPE_ID
    // }

    /// Returns `true` if the type id is [`I32`].
    ///
    /// [`I32`]: TypeId::I32
    #[must_use]
    pub fn is_i32(&self) -> bool {
        matches!(self, Self::I32)
    }

    /// Returns `true` if the type id is [`Void`].
    ///
    /// [`Void`]: TypeId::Void
    #[must_use]
    pub fn is_void(&self) -> bool {
        matches!(self, Self::Void)
    }

    /// Returns `true` if the type id is [`Unknown`].
    ///
    /// [`Unknown`]: TypeId::Unknown
    #[must_use]
    pub fn is_unknown(&self) -> bool {
        matches!(self, Self::Unknown)
    }
}

impl fmt::Display for TypeId {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            TypeId::Bool => write!(f, "bool"),
            TypeId::I32 => write!(f, "i32"),
            TypeId::Str => write!(f, "str"),
            TypeId::Void => write!(f, "void"),
            TypeId::Never => write!(f, "!"),
            TypeId::Unknown => write!(f, "unknown"),
            TypeId::Other(id) => write!(f, "[{id}]"),
        }
    }
}

//

fn skip_comments(token: &lexer::Result<SpannedToken>) -> bool {
    let Ok(token) = token else { return true };
    token.token() != Token::LineComment
}

//

#[derive(Debug)]
pub enum ParseStream<'a> {
    Lexer {
        lexer: Filter<Lexer<'a>, fn(&lexer::Result<SpannedToken>) -> bool>,
        peek: VecDeque<SpannedToken<'a>>,
    },
    Buffer {
        buffer: &'a [SpannedToken<'a>],
    },
}

impl<'a> ParseStream<'a> {
    const EOI: SpannedToken<'static> = SpannedToken::from_whole_str(Token::Eoi, "");

    pub const fn new(buffer: &'a [SpannedToken<'a>]) -> Self {
        Self::Buffer { buffer }
    }

    pub fn from_lexer(lexer: Lexer<'a>) -> Self {
        Self::Lexer {
            lexer: lexer.filter(skip_comments as _),
            peek: VecDeque::new(),
        }
    }

    pub fn look1(&mut self) -> Look {
        Look {
            token: self.top1().copied().unwrap_or(Self::EOI),
            arr: [Token::Eoi; 8],
            len: 0,
        }
    }

    pub fn look2(&mut self) -> Look {
        Look {
            token: self.top2().copied().unwrap_or(Self::EOI),
            arr: [Token::Eoi; 8],
            len: 0,
        }
    }

    pub fn peek1(&mut self, token: Token) -> bool {
        Some(token) == self.top1().map(|tok| tok.token())
    }

    pub fn peek2(&mut self, token: Token) -> bool {
        Some(token) == self.top1().map(|tok| tok.token())
    }

    pub fn next_token(&mut self) -> Result<SpannedToken<'a>> {
        Ok(self
            .next()
            .transpose()?
            .unwrap_or_else(|| SpannedToken::from_whole_str(Token::Eoi, "")))
    }

    pub fn expect_next(&mut self, token: Token) -> Result<SpannedToken<'a>> {
        let tok = self.next_token()?;
        if tok.token() == token {
            Ok(tok)
        } else {
            Err(unexpected(tok, &[token], false))
        }
    }

    fn top1(&mut self) -> Option<&'_ SpannedToken<'a>> {
        match self {
            ParseStream::Lexer { lexer, peek } => {
                if peek.is_empty() {
                    let next = lexer.next()?.unwrap_or(Self::EOI);
                    peek.push_back(next);
                }

                peek.front()
            }
            ParseStream::Buffer { buffer } => buffer.first(),
        }
    }

    fn top2(&mut self) -> Option<&'_ SpannedToken<'a>> {
        match self {
            ParseStream::Lexer { lexer, peek } => {
                if peek.len() < 2 {
                    let next = lexer.next()?.unwrap_or(Self::EOI);
                    peek.push_back(next);
                }

                if peek.len() == 1 {
                    let next = lexer.next()?.unwrap_or(Self::EOI);
                    peek.push_back(next);
                }

                peek.get(1)
            }
            ParseStream::Buffer { buffer } => buffer.first(),
        }
    }

    #[allow(clippy::let_and_return)]
    pub fn parse<T: Parse>(&mut self) -> Result<T> {
        // println!(
        //     "parsing {} (top: [{}, {}])",
        //     core::any::type_name::<T>(),
        //     self.top1().copied().unwrap_or(Self::EOI),
        //     self.top2().copied().unwrap_or(Self::EOI),
        // );

        let res = T::parse(self);

        // println!(
        //     " result {} = {:?}",
        //     core::any::type_name::<T>(),
        //     res.as_ref().map(|_| {})
        // );
        res
    }
}

impl<'a> Iterator for ParseStream<'a> {
    type Item = Result<SpannedToken<'a>>;

    fn next(&mut self) -> Option<Self::Item> {
        match self {
            ParseStream::Lexer { lexer, peek } => {
                if let Some(next) = peek.remove(0) {
                    return Some(Ok(next));
                }

                let res = lexer.next()?;
                Some(res.map_err(Error::Lexer))
            }
            ParseStream::Buffer { buffer } => {
                let (first, others) = buffer.split_first()?;
                *buffer = others;
                Some(Ok(*first))
            }
        }
    }
}

//

pub struct Look<'a> {
    token: SpannedToken<'a>,

    // expected one of: (overflow: ...)
    arr: [Token; 8],
    len: u8,
}

impl<'a> Look<'a> {
    pub fn peek(&mut self, expected: Token) -> bool {
        if self.token.token() == expected {
            return true;
        }

        if let Some(last) = self.arr.get_mut(self.len as usize) {
            *last = expected;
        }
        self.len += 1;

        false
    }

    pub fn err(self) -> Error {
        let arr = &self.arr[..self.arr.len().min(self.len as _)];
        let dots = self.len as usize > self.arr.len();

        unexpected(self.token, arr, dots)
    }
}

//

pub trait SingleToken: Sized {
    const TOKEN: Token;
    const SELF: Self;

    fn parse_single(tokens: &mut ParseStream) -> Result<Self> {
        let tok = tokens.next_token()?;
        if tok.token() != Self::TOKEN {
            return Err(unexpected(tok, &[Self::TOKEN], false));
        }

        Ok(Self::SELF)
    }
}

//

pub trait Parse: Sized {
    fn parse(tokens: &mut ParseStream) -> Result<Self>;
}

impl<T: Parse> Parse for Box<T> {
    fn parse(tokens: &mut ParseStream) -> Result<Self> {
        Ok(Box::new(tokens.parse()?))
    }
}

impl<T: Parse + SingleToken> Parse for Option<T> {
    fn parse(tokens: &mut ParseStream) -> Result<Self> {
        if tokens.peek1(T::TOKEN) {
            _ = tokens.next();

            Ok(Some(T::SELF))
        } else {
            Ok(None)
        }
    }
}

// impl<T: Parse> Parse for Vec<T> {
//     fn parse(tokens: &mut ParseStream) -> Result<Self> {
//         let mut result = Vec::new();

//         while let Some(next) = tokens.parse::<Option<T>>()? {
//             result.push(next);
//         }

//         Ok(result)
//     }
// }

//

pub fn unexpected(token: SpannedToken, arr: &[Token], dots: bool) -> Error {
    use std::fmt::Write;

    // let unexpected = unexpected
    //     .token()
    //     .as_token_str()
    //     .map(str::to_string)
    //     .unwrap_or_else(|| {
    //         let mut s = format!("{unexpected:?}");
    //         s.make_ascii_lowercase();
    //         s
    //     });

    let mut output = String::new();

    write!(&mut output, "`{token}`, expected ").unwrap();

    match arr {
        [] => panic!("expected nothing"),
        [only] => write!(&mut output, "`{only}`").unwrap(),
        [slice @ .., last] => {
            output.push_str("one of `");
            for token in slice {
                write!(&mut output, "{token}, ").unwrap();
            }
            write!(&mut output, "{last}").unwrap();

            if dots {
                output.push_str(", ...");
            }
            output.push('`');
        }
    }

    Error::UnexpectedToken(output)
}

//

#[cfg(test)]
mod tests {
    use lexer::Lexer;
    use serde::Serialize;

    use crate::{ast, Parse, ParseStream, Result};

    use insta::assert_yaml_snapshot;

    //

    fn parse<'a, T: Parse + Serialize + 'a>(str: &'a str) -> impl Serialize + 'a {
        let mut stream = ParseStream::from_lexer(Lexer::new(str));
        let ast: Result<ast::Ast<T>, _> = Parse::parse(&mut stream).map_err(|err| err.to_string());
        let extra = stream
            .next()
            .transpose()
            .map(|t| t.map(|t| t.as_str().to_string()))
            .map_err(|err| err.to_string());

        (str, ast, extra)
    }

    #[test]
    fn parse_test() {
        assert_yaml_snapshot!(parse::<ast::token::Comma>(","));
        assert_yaml_snapshot!(parse::<ast::Root>("val := "));
    }
}
