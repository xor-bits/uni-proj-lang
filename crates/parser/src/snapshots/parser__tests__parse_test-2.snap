---
source: crates/parser/src/lib.rs
expression: "parse::<ast::Root>(\"val := \")"
---
- "val := "
- Err: "unexpected token `<eoi>`, expected one of `{, LitInt, LitStr, Ident, (, fn`"
- Ok: ""

